pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +86,450 @@ const (
+		// controller, which is what we're ultimately trying to guard from
+		// downtime or excessive capacity reduction. This simplifies cases where
+		// old/new are asymmetrical and keeps things predictable.
+		percent := *config.UpdatePercent
+		maxIncrement = int(math.Ceil(float64(original) * (math.Abs(float64(percent)) / 100)))
+		// A negative percentage indicates our scale direction should be
+		// down/up instead of the default up/down.
+		if percent < 0 {
+			scaleUpFirst = false
+		}
+	}
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain available for the duration of the update if progress is to be
+	// made. If this is an up/down scale, impose a floor of 1 to avoid taking
+	// downtime. If the minimum available is >0 and the new controller hasn't
+	// yet been scaled up in a down/up update, it's possible that we can't
+	// progress without scaling up first (violating the direction) or scaling
+	// down first to zero (incurrent downtime). Rather than error out in that
+	// case, allow a minimum of zero. Users attempting a rolling update under
+	// those constraints with a down/up direction should be made aware that
+	// downtime is unavoidable.
+	minAvailable := original - maxIncrement
+	if minAvailable <= 0 {
+		if config.UpdatePercent == nil || (*config.UpdatePercent > 0) {
+			minAvailable = 1
+		} else {
+			minAvailable = 0
+		}
+	}
+	// Helpful output about what we're about to do.
+	direction := "down"
+	if scaleUpFirst {
+		direction = "up"
+	}
+	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval, maintain at least %d available)\n",
+		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, maxIncrement, minAvailable)
+
+	// Scale newRc and oldRc until newRc has the desired number of replicas and
+	// oldRc has 0 replicas.
+	increment := maxIncrement
+	progressDeadline := time.Now().UnixNano() + config.Timeout.Nanoseconds()
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		// Store the existing replica counts for progress timeout tracking.
+		newReplicas := newRc.Spec.Replicas
+		oldReplicas := oldRc.Spec.Replicas
+
+		if scaleUpFirst {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			scaleUpFirst = false
+		} // Otherwise, scale down/up.
+
+		// Scale down as much as possible while maintaining the minimum available
+		// amount. The scale-down informs us as to how much we can safely scale up
+		// next interval.
+		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minAvailable, config)
+		if err != nil {
+			return err
+		}
+		oldRc = scaledRc
+		increment = newIncrement
+
+		// Wait between down/up.
+		time.Sleep(config.UpdatePeriod)
```

Comment:
This is required, yes

On Thu, Aug 6, 2015 at 3:40 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/kubectl/rolling_updater.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11942#discussion_r36388371>
> :
>
> > +			newRc = scaledRc
> > +			scaleUpFirst = false
> > +		} // Otherwise, scale down/up.
> > +
> > +		// Scale down as much as possible while maintaining the minimum available
> > +		// amount. The scale-down informs us as to how much we can safely scale up
> > +		// next interval.
> > +		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minAvailable, config)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		oldRc = scaledRc
> > +		increment = newIncrement
> > +
> > +		// Wait between down/up.
> > +		time.Sleep(config.UpdatePeriod)
>
> So we decided to keep this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11942/files#r36388371>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift